### PR TITLE
Merge toolchain-2.9 into master and discard the toolchain-2.x branches?

### PR DESCRIPTION
**Pull request by meyerj**
> Who in maintainer of the toolchain branches? I would like this change in the `toolchain-2.9` branch.

I tried to keep them up-to-date in the past, but without actually using typelib (or [orogen](https://github.com/orocos-toolchain/orogen)) myself in the last years. I also do not know about gccxml vs. castxml.

Ideally the extra branches would not even be needed. The only non-trivial differences are in [`manifest.xml`](https://github.com/orocos-toolchain/typelib/compare/master...toolchain-2.9#diff-06e6215f32570c3a47a6a5336751fce3691dec5f07093b0f5214a65e07a49a1e) and [`package.xml`](https://github.com/orocos-toolchain/typelib/compare/master...toolchain-2.9#diff-37a67ff78eb7260214b323353263b9af40a2fa98719a1e81937fae0159df87a3), and related to the old problem that the same filename is used by [AutoProj](https://www.rock-robotics.org/documentation/autoproj/index.html) and ROS tools like [rosdep](https://wiki.ros.org/rosdep). The latter still prefers `manifest.xml` (dry) over `package.xml` (wet) if it exists, and handles `<depend package="...">` (other ROS packages) and `<rosdep name="...">` (packages provided via apt, pip, gems etc.) differently. That is unfortunate, the author of the package should not have to care and it depends on the build environment, which is probably why the difference has been removed for `package.xml` in [REP-0127](https://www.ros.org/reps/rep-0127.html) (latest specification is [REP-0149](https://ros.org/reps/rep-0149.html)). Additionally, AutoProj introduced useful new features like `<depend_optional>`, but which are not understood by ROS tools and cause errors. And last but not least, all the keys need to be either packages with a `package.xml` file that can be added to the workspace and built from source, or listed in the rosdep database maintained at https://github.com/ros/rosdistro/tree/master/rosdep.

So overall some coordination is required for every dependency update, to not break the build for either AutoProj or for ROS/catkin users.

The version numbers are indeed not so important, especially when building from source. But the version has to be increased for each new release into ROS - provided that this is still desirable. The last release of typelib was [into ROS indigo back in 2015](http://repositories.ros.org/status_page/compare_indigo_kinetic_lunar_melodic.html?q=typelib) ([orocos-gbp/typelib-release](https://github.com/orocos-gbp/typelib-release)). In any case the version numbers and release cycles should IMHO not be coupled to [RTT](https://github.com/orocos-toolchain/rtt) or [Orocos Toolchain](https://github.com/orocos-toolchain/orocos_toolchain) anymore (and hence no `toolchain-2.x` branches). So please choose whatever you prefer for typelib and other packages you maintain (https://github.com/orocos-toolchain/orogen/pull/140#issuecomment-655501107).

Other changes ([master...toolchain-2.9](https://github.com/orocos-toolchain/typelib/compare/master...toolchain-2.9)) are actually trivial and could be either reverted or applied to the `master` branch?
- install `package.xml` in `CMakeLists.txt` and update the outdated maintainer email address `orocos-dev@lists.mech.kuleuven.be` (replaced by `orocos-dev@orocos.org`, which currently forwards to [orocosdev@ls.kuleuven.be](https://ls.kuleuven.be/cgi-bin/wa?A0=OROCOSDEV))
- rosbuild `Makefile` is obsolete and has been removed (https://github.com/orocos-toolchain/typelib/commit/5e8b7c8d2459bcee5f393d3039b8ae0ee7458716)
- fixes needed for linking to pthread and `Boost_THREAD_LIBRARIES` (https://github.com/orocos-toolchain/typelib/commit/40cd3a89efb4307db59324e8aabf8b7bc498bdf7)?
- some arch detection patch required for macOS back then, likely obsolete (https://github.com/orocos-toolchain/typelib/commit/6339de6629fa2c371c7f8a59fed7e17d71e2d558)?

See also https://github.com/orocos-toolchain/orogen/pull/140 and https://github.com/orocos-toolchain/rtt/pull/321. It would be very nice if we can find a solution for those kind of problems in the future, but I do not see how this can be done without deprecating `manifest.xml` in favor of `autoproj.xml` and `package.xml` (https://github.com/orocos-toolchain/rtt/pull/321#issuecomment-631755015), and keeping both up-to-date? If there is anything I can do (other than merging the two approved pull requests above), please let me know.

_Originally posted by @meyerj in https://github.com/orocos-toolchain/typelib/issues/133#issuecomment-766974931_

### CODE REVIEW

File: typelib/CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +15,7 @@ endif (NOT TYPELIB_HARDCODED_PLUGIN_PATH)
 set_source_files_properties(endianness.cc PROPERTIES
     COMPILE_FLAGS -fno-strict-aliasing)
 
-target_link_libraries(typeLib ${CMAKE_DL_LIBS} ${Boost_FILESYSTEM_LIBRARIES} ${Boost_SYSTEM_LIBRARIES})
+target_link_libraries(typeLib pthread ${CMAKE_DL_LIBS} ${Boost_FILESYSTEM_LIBRARIES} ${Boost_SYSTEM_LIBRARIES} ${Boost_THREAD_LIBRARIES})
```
Comment (meyerj) [inline typelib/CMakeLists.txt:18]: Boost mutexes are used in [typelib/utilmm/singleton/server.hh](https://github.com/orocos-toolchain/typelib/blob/69476e5e886138d1626dc234a38198d89d5063cb/typelib/utilmm/singleton/server.hh#L9) and hence the library must be linked to [Boost Thread](https://www.boost.org/doc/libs/1_75_0/doc/html/thread.html). Could be that `boost::recursive_mutex` is header-only if compiled in Release mode, but we observed linker errors in the past, at least in `Debug` mode.

File: manifest.xml [hunk 2]
```diff
@@ -23,12 +23,12 @@
...
-  <depend package="libxml2" />
-  <depend package="pkg-config" />
-  <depend package="ruby-dev" />
-  <depend package="ruby-backports" />
+  <rosdep name="boost" />
+  <rosdep name="facets" />
+  <rosdep name="libxml2" />
+  <rosdep name="pkg-config" />
+  <rosdep name="ruby-dev" />
+  <rosdep name="ruby-backports" />
```
Comment (meyerj) [inline manifest.xml:31]: This is required to find the matching keys in the [rosdep database](https://github.com/ros/rosdistro/tree/master/rosdep).
Alternatively the file could be renamed to `autoproj.xml` in the future, if that is an acceptable path forward for you.

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 3]
```diff
@@ -1000,6 +1000,11 @@ def self.gccxml(file, options)
             if raw = options[:rawflags]
                 cmdline.concat(raw)
             end
+            if 1.size == 4
+                cmdline << "-m32"
+            else
+                cmdline << "-m64"
+            end
```
Comment (meyerj) [inline bindings/ruby/lib/typelib/gccxml.rb:1007]: Likely can be discarded if gccxml is obsolete anyway.

File: CMakeLists.txt [hunk 4]
```diff
@@ -62,12 +62,13 @@ ENDIF(NOT LibXML_FOUND)
 
 CONFIGURE_FILE(typelib.pc.in typelib.pc @ONLY)
 INSTALL(FILES ${CMAKE_BINARY_DIR}/typelib.pc DESTINATION lib/pkgconfig)
+INSTALL(FILES package.xml DESTINATION share/typelib)
```
Comment (meyerj) [inline CMakeLists.txt:65]: Installation of `package.xml` is required to make ROS tools happy.

**Comment by doudou** [comment]
I don't see any blocker in there. @g-arjones ?

**Comment by doudou** [comment]
About the `autoproj.xml` file, we can definitely implement it ... I have to admit it's low on my priority list right now.

What confuses me is:

If **you** are not using typelib/orogen, doesn't it mean that the whole integration with the rest of your orocos/rtt/ros toolchain is falling behind ? So why not deprecate it once and for all and call it a day ?

**Comment by meyerj** [comment]
> If **you** are not using typelib/orogen, doesn't it mean that the whole integration with the rest of your orocos/rtt/ros toolchain is falling behind ? So why not deprecate it once and for all and call it a day ?

That is what I already suggested earlier, for example in https://github.com/orocos-toolchain/orocos_toolchain/pull/37#issuecomment-631124521. So basically to not consider all packages together as "the Orocos Toolchain" anymore. That is de-facto the case, and only very few changes in one toolchain package actually require synchronized changes in another. So better to maintain, version and release them as independent projects, where [orogen](https://github.com/orocos-toolchain/orogen) and [rtt_typelib](https://github.com/orocos-toolchain/rtt_typelib) just depend on (ideally) any recent version of RTT.

If there is agreement to go down that path, then basically the [orocos_toolchain](https://github.com/orocos-toolchain/orocos_toolchain) repository would reach its end-of-life, or only provide [log4cpp](https://github.com/orocos-toolchain/log4cpp), [rtt](https://github.com/orocos-toolchain/rtt) and [ocl](https://github.com/orocos-toolchain/ocl). If that is purely my choice, then so be it, purely for practical reasons and without that I want to offend anyone. Furthermore all macros related to `typegen` could be removed from [UseOROCOS-RTT.cmake](https://github.com/orocos-toolchain/rtt/blob/master/UseOROCOS-RTT.cmake) at some point, like we also removed most traces of ROS integration in core RTT, except for build tool support. If needed, orogen can provide CMake macros itself after `find_package(orogen)`. To be decided whether that also implies moving repositories [rtt](https://github.com/orocos-toolchain/rtt) and [ocl](https://github.com/orocos-toolchain/ocl) to the GitHub [orocos](https://github.com/orocos) organization and other repositories to other organizations, for example [rock-core](https://github.com/rock-core).

But obviously there are some users who want to build typelib and orogen with ROS tools like catkin (https://github.com/orocos-toolchain/typelib/pull/133), and that might be worth to support independent of Orocos? Actually releasing packages into ROS, such that the buildfarm builds and provides Debian, Ubuntu and Fedora packages, is optional and would put some restrictions on the versions of Ruby and other dependencies that can be used (following the [ROS target platforms](https://www.ros.org/reps/rep-0003.html)). But whether to support source builds with ROS tools or not is then up to you or another maintainer to decide. Basically the only requirements are a `CMakeLists.txt` and an up-to-date `package.xml` file, and to somehow solve the `manifest.xml` incompatibility issues.

**Comment by g-arjones** [comment]
> I don't see any blocker in there. @g-arjones ?

Nope, looks good to me as well.

**Review by doudou — APPROVED** [review]

**Pull request merged**